# <span style="color:rgb(255, 0, 255)"> Lab Cross Validation
<span style="color:rgb(255, 0, 255)"> Ainara Guerra
    

**But first, libraries:**

In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import matplotlib.ticker as mk
pd.set_option('display.max_columns', None)

import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

from sklearn.preprocessing import Normalizer
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from sklearn.linear_model import LogisticRegression

from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

import os #we will use the function listdir to list files in a folder
import math #to apply absolute value

### <span style="color:rgb(255, 0, 255)">--- Load the database</span>

In [2]:
data = pd.read_csv('Customer-Churn.csv')

### <span style="color:rgb(255, 0, 255)">--- Let's look at its main features (head, shape, info).

In [3]:
data.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female              0     Yes         No       1           No   
1    Male              0      No         No      34          Yes   
2    Male              0      No         No       2          Yes   
3    Male              0      No         No      45           No   
4  Female              0      No         No       2          Yes   

  OnlineSecurity OnlineBackup DeviceProtection TechSupport StreamingTV  \
0             No          Yes               No          No          No   
1            Yes           No              Yes          No          No   
2            Yes          Yes               No          No          No   
3            Yes           No              Yes         Yes          No   
4             No           No               No          No          No   

  StreamingMovies        Contract  MonthlyCharges TotalCharges Churn  
0              No  Month-to-month           29.85        29.85    No  
1              No        One year           56.95       1889.5    No  
2              No  Month-to-month           53.85       108.15   Yes  
3              No        One year           42.30      1840.75    No  
4              No  Month-to-month           70.70       151.65   Yes

In [4]:
data.shape
# It has 7043 rows x 16 columns, not so bad

(7043, 16)

In [5]:
data.info()
# Here we can see it apparently doesn't have nulls, but we should deep into each column to see how their values are distributed.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7043 non-null   object 
 1   SeniorCitizen     7043 non-null   int64  
 2   Partner           7043 non-null   object 
 3   Dependents        7043 non-null   object 
 4   tenure            7043 non-null   int64  
 5   PhoneService      7043 non-null   object 
 6   OnlineSecurity    7043 non-null   object 
 7   OnlineBackup      7043 non-null   object 
 8   DeviceProtection  7043 non-null   object 
 9   TechSupport       7043 non-null   object 
 10  StreamingTV       7043 non-null   object 
 11  StreamingMovies   7043 non-null   object 
 12  Contract          7043 non-null   object 
 13  MonthlyCharges    7043 non-null   float64
 14  TotalCharges      7043 non-null   object 
 15  Churn             7043 non-null   object 
dtypes: float64(1), int64(2), object(13)
memory

In [6]:
data.isna().sum()
# another way of see the nulls

gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

***

### <span style="color:rgb(255, 0, 255)">--- Code from previous lab starts here

In [8]:
# I took some of the code that might be interesting from the previous lab to use it in this also.

#### <span style="color:rgb(255, 0, 255)"> ---- **Read that data into Python and call the dataframe churnData:**

In [9]:
ChurnData = data.copy()

#### <span style="color:rgb(255, 0, 255)"> ----  Check the datatypes of all the columns in the data. You would see that the column TotalCharges is object type. Convert this column into numeric type using pd.to_numeric function:

In [10]:
#Checking dtypes of the dataset:
ChurnData.dtypes

gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

In [11]:
ChurnData["TotalCharges"]
# It is indeed a number.

0         29.85
1        1889.5
2        108.15
3       1840.75
4        151.65
         ...   
7038     1990.5
7039     7362.9
7040     346.45
7041      306.6
7042     6844.5
Name: TotalCharges, Length: 7043, dtype: object

In [12]:
# Let's explore it before turning into a numeric
ChurnData["TotalCharges"].value_counts()


          11
20.2      11
19.75      9
20.05      8
19.9       8
          ..
6849.4     1
692.35     1
130.15     1
3211.9     1
6844.5     1
Name: TotalCharges, Length: 6531, dtype: int64

In [13]:
# We see there are 11 values with empty spaces. Looks good to me to turn it into zero in we don't have value for that. 
#It could also be a good idea to drop them since we don't have much
ChurnData["TotalCharges"].replace(" ", "0", inplace=True)

In [14]:
ChurnData["TotalCharges"].value_counts()
# It worked

0         11
20.2      11
19.75      9
20.05      8
19.9       8
          ..
6849.4     1
692.35     1
130.15     1
3211.9     1
6844.5     1
Name: TotalCharges, Length: 6531, dtype: int64

In [15]:
# Now er can convert it to numeric:
ChurnData["TotalCharges"] = pd.to_numeric(ChurnData["TotalCharges"], errors="coerce")

In [16]:
# let's see if it worked by checking the NaNs:
ChurnData["TotalCharges"].isna().sum()
#It worked

0

#### <span style="color:rgb(255, 0, 255)"> ---- Check for null values in the dataframe. Replace the null values.

In [17]:
# We saw earlier that there weren't NaNs per se. 
# Let's check value counts of each column to see if theres anything we need to tackle (detective mode on🔍)

# 1. let's remember the columns:
ChurnData.columns

Index(['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure',
       'PhoneService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [18]:
for col in ChurnData.columns:
    print(ChurnData[col].value_counts())

Male      3555
Female    3488
Name: gender, dtype: int64
0    5901
1    1142
Name: SeniorCitizen, dtype: int64
No     3641
Yes    3402
Name: Partner, dtype: int64
No     4933
Yes    2110
Name: Dependents, dtype: int64
1     613
72    362
2     238
3     200
4     176
     ... 
28     57
39     56
44     51
36     50
0      11
Name: tenure, Length: 73, dtype: int64
Yes    6361
No      682
Name: PhoneService, dtype: int64
No                     3498
Yes                    2019
No internet service    1526
Name: OnlineSecurity, dtype: int64
No                     3088
Yes                    2429
No internet service    1526
Name: OnlineBackup, dtype: int64
No                     3095
Yes                    2422
No internet service    1526
Name: DeviceProtection, dtype: int64
No                     3473
Yes                    2044
No internet service    1526
Name: TechSupport, dtype: int64
No                     2810
Yes                    2707
No internet service    1526
Name: StreamingTV, 

In [19]:
# This looks good to me, but what if we have some values = " ", as we had in TotalCharges that weren't spotted?
# Let's discover if there is anything else:
empty_columns = (ChurnData == " ").sum()

# Iterate over the columns and print the count of empty values per column
for column, count in empty_columns.iteritems():
    if count > 0:
        print(f"Column '{column}' has {count} empty value(s).")
    if empty_columns.sum() == 0:
        print(f"Column '{column}' has not empty values.")
        
#Apparently, looks clean to me. 

Column 'gender' has not empty values.
Column 'SeniorCitizen' has not empty values.
Column 'Partner' has not empty values.
Column 'Dependents' has not empty values.
Column 'tenure' has not empty values.
Column 'PhoneService' has not empty values.
Column 'OnlineSecurity' has not empty values.
Column 'OnlineBackup' has not empty values.
Column 'DeviceProtection' has not empty values.
Column 'TechSupport' has not empty values.
Column 'StreamingTV' has not empty values.
Column 'StreamingMovies' has not empty values.
Column 'Contract' has not empty values.
Column 'MonthlyCharges' has not empty values.
Column 'TotalCharges' has not empty values.
Column 'Churn' has not empty values.


#### <span style="color:rgb(255, 0, 255)"> ---- Use the following features: tenure, SeniorCitizen, MonthlyCharges and TotalCharges:
    
1. Scale the features either by using normalizer or a standard scaler.
2. Split the data into a training set and a test set.
3. Fit a logistic regression model on the training data.
4. Check the accuracy on the test data.

In [20]:
data_df = ChurnData[["tenure", "SeniorCitizen", "MonthlyCharges", "TotalCharges"]]
data_df

tenure  SeniorCitizen  MonthlyCharges  TotalCharges
0          1              0           29.85         29.85
1         34              0           56.95       1889.50
2          2              0           53.85        108.15
3         45              0           42.30       1840.75
4          2              0           70.70        151.65
...      ...            ...             ...           ...
7038      24              0           84.80       1990.50
7039      72              0          103.20       7362.90
7040      11              0           29.60        346.45
7041       4              1           74.40        306.60
7042      66              0          105.65       6844.50

[7043 rows x 4 columns]

In [21]:
# 1. Scale the features either by using normalizer or a standard scaler.
transformer = MinMaxScaler().fit(data_df)
data_df_minmax = transformer.transform(data_df)
data_df_norm = pd.DataFrame(data_df_minmax,columns=data_df.columns)
data_df_normalized = data_df_norm.copy()
data_df_normalized.shape

(7043, 4)

## ML without upsampling or downsampling

In [22]:
# 2. Split the data into a training set and a test set.
X = data_df_normalized
y = ChurnData["Churn"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42) 

#### Logistic Regression

In [23]:
# Create an instance of Logistic Regression
logreg = LogisticRegression()

# Train the model using the training data
logreg.fit(X_train, y_train)

# Use the trained model to make predictions on the test data
y_pred = logreg.predict(X_test)

# Evaluate the model
accuracy = logreg.score(X_test, y_test)

# Print the accuracy
print("Accuracy:", accuracy)

Accuracy: 0.7989778534923339


In [42]:
#This is the result of logaristic regression without any downsampling or upsampling
# Let's dive also into the results of the Decission Tree Classifier

#### Decission Tree Classifier

In [26]:
# Create an instance of Decision Tree Classifier
classifier = DecisionTreeClassifier()

# Train the model using the training data
classifier.fit(X_train, y_train)

# Use the trained model to make predictions on the test data
y_pred = classifier.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)

# Print the accuracy
print("Accuracy:", accuracy)

Accuracy: 0.7075525269733106


### 3. Synthetic Minority Oversampling TEchnique (SMOTE)

#### Logistic Regression

In [27]:
from imblearn.over_sampling import SMOTE
smote = SMOTE()
X = ChurnData[['tenure', 'SeniorCitizen','MonthlyCharges']]
transformer = StandardScaler().fit(X)
X = transformer.transform(X)
y = ChurnData['Churn']
X_sm, y_sm = smote.fit_resample(X, y)
y_sm.value_counts()

No     5174
Yes    5174
Name: Churn, dtype: int64

In [28]:
X_train, X_test, y_train, y_test = train_test_split(X_sm, y_sm, test_size=0.3, random_state=100)
classification = LogisticRegression(random_state=0, multi_class='ovr').fit(X_train, y_train)
predictions = classification.predict(X_test)

classification.score(X_test, y_test)

0.743317230273752

#### Decission Tree Classifier

In [29]:
# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X_sm, y_sm, test_size=0.3, random_state=100)

# Create an instance of Decision Tree Classifier
classifier = DecisionTreeClassifier(random_state=0)

# Train the model using the training data
classifier.fit(X_train, y_train)

# Use the trained model to make predictions on the test data
predictions = classifier.predict(X_test)

# Calculate the accuracy score
accuracy = accuracy_score(y_test, predictions)

# Print the accuracy score
print("Accuracy:", accuracy)

Accuracy: 0.7600644122383253


### 4. UnderSampling using TomekLinks 

In [30]:
from imblearn.under_sampling import TomekLinks

tl = TomekLinks()
X_tl, y_tl = tl.fit_resample(X, y)
y_tl.value_counts()

No     4697
Yes    1869
Name: Churn, dtype: int64

In [31]:
X_tl2, y_tl2 = tl.fit_resample(X_tl, y_tl)
y_tl2.value_counts()

No     4544
Yes    1869
Name: Churn, dtype: int64

In [32]:
# As we learnt in the class, it does not make the two classes equal but only removes the points from the majority class that are close to other points in minority class

In [33]:
X_train, X_test, y_train, y_test = train_test_split(X_tl, y_tl, test_size=0.3, random_state=100)
classification = LogisticRegression(random_state=0, multi_class='ovr').fit(X_train, y_train)
predictions = classification.predict(X_test)

classification.score(X_test, y_test)

0.8040609137055837

In [38]:
# Split the data into training and test sets
X_train, X_test, y_train, y_test_r = train_test_split(X_tl, y_tl, test_size=0.3, random_state=100)

# Create an instance of Decision Tree Classifier
classifier = DecisionTreeClassifier(random_state=0)

# Train the model using the training data
classifier.fit(X_train, y_train)

# Use the trained model to make predictions on the test data
y_pred = classifier.predict(X_test)

# Calculate the accuracy score
accuracy = accuracy_score(y_test_r, y_pred)

# Print the accuracy score
print("Accuracy:", accuracy)

Accuracy: 0.7644670050761422


### Conclusions:

**Let's recall the results of accuracy first**
- ML without upsampling or downsampling
        - Logistic Regression: 0.798
        - Decission Tree Classifier: 0.707
- SMOTE
        - Logistic Regression. 0.743
        - Decission Tree Classifier: 0.760
- TomekLinks
        - Logistic Regression: 0.804
        - Decission Tree Classifier: 0.764

**1.** Decission Tree Classifier is sensitive to SMOTE and TomekLinks sampling methods. The second one is the best for this prediction model. 

**2.** Nevertheless, this sampling  methods don't do any significant change to the results of logistic regression accuracy. 

<!-- <span style="color:rgb(255, 0, 255)"> --- **Read that data into Python and call the dataframe churnData:** -->